python/mlx/optimizers/optimizers.py

```diff
@@ -736,3 +736,17 @@ def apply_single(self, gradient: mx.array, parameter: mx.array, state: dict):
         if self.weight_decay != 0:
             parameter += parameter * (-self.weight_decay * learning_rate)
         return parameter - update
+
+
+def clip_grad_norm(grads, max_norm):
+    norm_squared = tree_reduce(
+        lambda acc, g: acc + g.square().sum(), grads, mx.array(0.0)
```

Comment:
I would use `0.0` here instead of `mx.array(0.0)` then it will be weakly typed so we don't need to worry about inadvertent up-casting.

Reply:
I guess you could also leave it unspecified and use the first leaf?